src/main/java/org/jabref/gui/actions/WriteXMPAction.java

```diff
@@ -1,20 +1,266 @@
+
+            if (files.isEmpty()) {
+                skipped++;
+                SwingUtilities.invokeLater(() -> optDiag.getProgressArea()
+                                                        .append("  " + Localization.lang("Skipped - No PDF linked") + ".\n"));
+            } else {
+                for (Path file : files) {
+                    if (Files.exists(file)) {
+                        try {
+                            XmpUtilWriter.writeXmp(file, entry, database, Globals.prefs.getXMPPreferences());
+                            SwingUtilities.invokeLater(
+                                    () -> optDiag.getProgressArea().append("  " + Localization.lang("OK") + ".\n"));
+                            entriesChanged++;
+                        } catch (Exception e) {
+                            SwingUtilities.invokeLater(() -> {
+                                optDiag.getProgressArea().append("  " + Localization.lang("Error while writing") + " '"
+                                        + file.toString() + "':\n");
+                                optDiag.getProgressArea().append("    " + e.getLocalizedMessage() + "\n");
+                            });
+                            errors++;
+                        }
+                    } else {
+                        skipped++;
+                        SwingUtilities.invokeLater(() -> {
+                            optDiag.getProgressArea()
+                                   .append("  " + Localization.lang("Skipped - PDF does not exist") + ":\n");
+                            optDiag.getProgressArea().append("    " + file.toString() + "\n");
+                        });
+                    }
+                }
+            }
+
+            if (optDiag.isCanceled()) {
+                SwingUtilities.invokeLater(
+                        () -> optDiag.getProgressArea().append("\n" + Localization.lang("Operation canceled.") + "\n"));
+                break;
+            }
+        }
+        SwingUtilities.invokeLater(() -> {
+            optDiag.getProgressArea()
+                   .append("\n"
+                           + Localization.lang("Finished writing XMP for %0 file (%1 skipped, %2 errors).", String
+                           .valueOf(entriesChanged), String.valueOf(skipped), String.valueOf(errors)));
+            optDiag.done();
+        });
+
+        if (!goOn) {
+            return;
+        }
+
+        basePanel.output(Localization.lang("Finished writing XMP for %0 file (%1 skipped, %2 errors).",
+                String.valueOf(entriesChanged), String.valueOf(skipped), String.valueOf(errors)));
     }
 
+    class OptionsDialog extends JabRefDialog {
+
+        private final JButton okButton = new JButton(Localization.lang("OK"));
+        private final JButton cancelButton = new JButton(Localization.lang("Cancel"));
+
+        private boolean canceled;
```

Comment:
Also here the `canceled` could be named more precisely: what should be canceled?